scheduler/src/cook/compute_cluster.clj

```diff
@@ -43,7 +45,12 @@
 ;
 ; So, we must grab this lock before calling kill-task in the compute cluster API. As all of our invocations to it are via
 ; safe-kill-task, we add the lock there.
-(def kill-lock-object (Object.))
+;
+; We use a ReaderWriterLock where launch tasks grab the readers, so multiple pools can launch at the same time, and a writer lock
+; here so that kills are blocked until nobody is in the middle of launching. Note that we don't actually need to hold this lock while
+; we kill, just to delay the kill until after any task-id written in the database is also guaranteed to be into the backend.
+;
+(def ^ReentrantReadWriteLock kill-lock-object (ReentrantReadWriteLock. true))
```

Comment:
So we still block when acquiring a write lock on one pool and trying to read on a different pool? If we had lock per-pool then we would actually decouple the pools?  We just don't want to refactor to pass in pool to kill-task?

Reply:
Yes and no. Since a compute cluster has multiple pools, we don't shard based on the pool. Yes, we do block, but the lock is only acquired for microseconds, so we block for microseconds for the kill write lock to be acquired. We're actually just serializing on this lock.